### ISSUE TITLE
Make downloaded zip self-documenting

### ISSUE THREAD

**Issue by phraenquex** [m2ms/fragalysis-frontend#747]
The zipfile needs to be thoroughly self-documenting. 

Include the following: 

### The URL that recreates this download
fragalysis-download-link.URL - text file containing the URL that is also returned in the "Copy Link" button of the modal

### The json that is sent to the backend
download-api-call.json - json string that is also available from the "Copy API json call" button on the modal.

### Documentation of the directory and file hierarchy
The old zip files had something - I think @TJGorrie provides it in the upload.
@TJGorrie please attach to this ticket the file that should now be included

### Declare what is the reference file(s)
Put all the references in an identifable format. (Is that what the "references.....pdb" are from #757 
Action: 
1. will be hacked/fixed in #757 for Mpro.
2. in future, will be part of upload Epic.

**Comment by duncanpeacock** [comment]
The first two points should be straightforward - I just need to confirm what the frontend does.

For the last two points, I'm not sure - need to confirm with Tyler I think.

**Comment by duncanpeacock** [comment]
This was discussed with Tyler on Monday afternoon.
ReadMe is attached:
[readme.md](https://github.com/m2ms/fragalysis-frontend/files/7940424/readme.md)

This will be kept in the Fragalysis Backend Repo and used as a basis for each download with the other above (URL/Json/Files) appended to the end. If you need to change the ReadMe test, we can simply update the version in the repo.

For the list of files in the download file, I suggest adding a table to the readme. There is a simple method in python to extract the file paths to get a list like:

Filepath
aligned/Mpro_1234_0A/Mpro_1234_0A.pdb
aligned/Mpro_1234_0A/Mpro_1234_0A.sdf

I can also easily add a size of file if that is useful.

**Comment by duncanpeacock** [comment]
Pull request: https://github.com/xchem/fragalysis-backend/pull/303

**Comment by phraenquex** [comment]
Has this been merged into staging?  Because there's no .pdf file in the .zip archive of this download:
https://fragalysis.xchem.diamond.ac.uk/viewer/react/download/tag/9da6b9ad-e3fa-4083-a8e2-a76c664ccf26
@duncanpeacock

**Comment by duncanpeacock** [comment]
Yes, we needed to bounce staging, sorry. It is there now - I just did a down load following Tim bouncing the stack. You'll need to either prepare a brand new search or wait an hour for the cache files to be regenerated to see the Readme..

**Comment by phraenquex** [comment]
Terrific!  Rearrange file, please (while we're at it):

# Documentation for the downloaded zipfile  [Top heading]
## Download details  [subheading]
### Download URLs  [sub-subheading]
_Download URL_: https://fragalysis.xchem.diamond.ac.uk/viewer/react/download/tag/a7ea4b13-90b2-4040-8396-6d3fe7b111a3
_Download snapshot_:  https://https://fragalysis.xchem.diamond.ac.uk/viewer/react/projects/1350/1010"
`[Ensure they render as clickable hyperlink in the pdf, and include https://]`

### Download options selected
The following options were checked in the download dialogue: 
* Selected: All structures
* Selected: PanDDA Event maps - primary evidence
* Selected: Incremental - always up-to-date with latest structures
* Selected: Single SDF of all ligands

`[use actual words in the modal - just type them across if necessary]`

### Download command (JSON)   [currently: "Original search"]
`[Add a brief description of what this means, e.g. ]`
JSON command sent from front-end to backend to generate the download.  This can be reused programmatically as a POST command
The actual JSON (as currently)

## Directory structure  [subheading] [note spelling mistake!]
`[The text as you currently have it.]`

## Files included
`[The list as you currently have it]`

**Comment by duncanpeacock** [comment]
Most of this is straightforward with the exception of the:
download snapshot: https://https://fragalysis.xchem.diamond.ac.uk/viewer/react/projects/1350/1010"

This is more complicated.

I currently don't have this information in the back end as the snapshot is created by the frontend (unless there is a really obvious connection I am not aware of).
 
We would need to change the API to add this link. The frontend will have to be changed to supply the URL to the API so that it can be added to the file.

But it also creates another issue. At the moment if you request the same file from different snapshots and the file is the same, then it can use the one in cache giving an instant response. 

To enable this to work I would either have:
1. To store the URL (meaning the caching won't work nearly as efficiently because we would need to keep different files for each snapshot URL) or 
2. I'd have to pick up the zip file each time (whether cached or not), generate the Readme with the URL and add it to it.  

2) Seems to be a more efficient  solution, but as I say it also requires a small front end change.

**Comment by duncanpeacock** [comment]
Actually, there is another issue here - looking at the detail.
I only know in the back end whether the user has selected all structures (which I think is a target level search) or not - otherwise I just get a list of proteins. I think that this will also have to be provided in the API call by the frontend.

So maybe for this, the best solution is a bit like 2) in my previous comment - but just accept back a (JSON or markdown?) string from the front end in the API call that I can add into the relevant sections. That also allows the front end to change without having to change the backend code - we just agree the section headings/format. I'd have to work out the best way to do this with Boris.

**Comment by phraenquex** [comment]
Can Boris stick the GUI checks into a string and post it back to Duncan in some text field?  Because the GUI options do not match the API options 1:1.  
I know it's hacky - but hacky is probably appropriate.
Of course, we could simply leave that bit out of the documentation - but I've decided it's probably really important, to have the zip self-documented, just because data downloads take on a life of their own.

@boriskovar-m2ms @duncanpeacock please give me a sanity check on how much additional work this entails.

**Comment by boriskovar-m2ms** [comment]
@phraenquex From the front-end point of view this is like 2-3 hours max

**Comment by duncanpeacock** [comment]
I've discussed this with Boris and I think I have a solution which will/should work.

These are the essential steps:

1. We add a JSON string to the POST fields in the API (called document_details). This will contain the information that the front end needs to supply to the back end (essentially the contents  of the Download details section above which is all really front end stuff) and will be filled in every time the file is requested (since the same zip contents can attached to more than one snapshot).
2. The back end will now create two separate files - the zip contents without the Readme (which will be cached etc) and the Readme.pdf (which is generated fresh every time based on the document_details JSON string - even if the file is persistent). 
3. The GET call will then add the Readme.pdf to a copy of the zip file and return the combined file. You can't modify a file in an existing zip file without rewriting the whole zip, but you can easily add a file to an existing zip - hence the solution. 

The complexity is in the JSON string, so it would be best if we can keep that quite simple. Any formatting of the PDF should be done by the backend, so the JSON needs to just indicate what section the data item belongs to and what it is. A structure like this should work:    

{ 
    "download_urls" : 
        {"download_url" : "/viewer/react/download/tag/etc...", 
         "snapshot_url" : "/viewer/react/projects/1350/1010"
     }, 
    "download_options_list" : {"Selected: All structures" : "True", etc.}, 
    "download_command_json": " .. " 
}

I've made the URLs relative as the back end should add the host/HTTPS when it adds the URLs to the file. 

In terms of work, it is a bit fiddly as I explained before - I need to construct the Readme using the JSON and I had a few problems with formatting and the pandoc library (that does the translation) first time round. 
So I think we're looking at around **8 - 10 hours work** with some joint testing with Boris to make sure the front and back ends are talking to each other properly before we go to staging.

**Comment by phraenquex** [comment]
Thanks, both. 
@duncanpeacock when you say "8-10 hours work _with some joint testing_" - do the 8-10 hours include the testing? 

Either way, please create a new ticket, with the spec.  (Copy-paste from above?)
Title:  "Finalise self-documentation of zip archive"

I assume the rest of the re-formatting was not an issue.

**Comment by phraenquex** [comment]
I've created the issue;  please copy across as relevant.

**Comment by duncanpeacock** [comment]
Relevant sections copied over

**Comment by phraenquex** [comment]
Please fix everything except what needs frontend info, that's covered in #785.

**Comment by duncanpeacock** [comment]
OK, I'll split #785 into two tickets.
